File: mahjong.cpp
//
// Created by 洪健淇 on 2019/10/30.
//

#include <iostream>
#include <vector>
#include <chrono>
#include <random>

using namespace std;

class Tile {
private:
    inline static const string strs[10] = {"萬", "條", "筒", "東", "南", "西", "北", "中", "發", "白"};
    static constexpr int nums[10] = {0, 20, 40, 60, 65, 70, 75, 80, 85, 90};
    inline static int allTiles[34] = {0};
    inline static int deckTiles[136] = {0};

    template<class T> static int getIndex(const T *arr, const T& item) {
        auto itr = find(arr, arr + 10, item);
        return distance(arr, itr);
    };

    static string numberToString(const int& num) {
        if (num >= 60) {
            return strs[getIndex(nums, num)];
        }
        int order = num % 10;
        return to_string(order) + strs[getIndex(nums, num - order)];
    }

    static int make(const string& type, const int& order=0) {
        return nums[getIndex(strs, type)] + order;
    }

    static vector<int>getDuplicates(const vector<int>& numbers, const int& times) {
        int last = numbers[0];
        int ocu = 1;
        vector<int> result;
        for (size_t i = 1; i < numbers.size(); i++) {
            int num = numbers[i];
            if (num == last) {
                ocu += 1;
                if (ocu == times) {
                    result.push_back(num);
                }
            } else {
                ocu = 1;
            }
            last = num;
        }
        return result;
    }

    static vector<int> removeDuplicates(const vector<int>& numbers, const int& delNum, int times) {
        vector<int> result;
        for (int num : numbers) {
            if (num == delNum && times != 0) {
                times -= 1;
                continue;
            }
            result.push_back(num);
        }
        return result;
    }

    static bool isResolved(const vector<int>& numbers) {
        bool isIgnore = false;
        int ignoreTimes = 0;

        for (size_t idx = 0; idx < numbers.size(); idx++) {
            if (isIgnore && ignoreTimes != 0) {
                ignoreTimes -= 1;
                if (ignoreTimes == 0) {
                    isIgnore = -1;
                }
                continue;
            }
            if (idx + 3 > numbers.size()) {
                return false;
            }

            int i = numbers[idx];
            if (i == numbers[idx + 2]) {
                isIgnore = true;
                ignoreTimes = 2;
                continue;
            }

            if (i < 60) {
                int step = 1;
                vector<int> lst;
                for (size_t ia = idx + 1; ia < numbers.size(); ia++) {
                    int a = numbers[ia];
                    if (step == -1) {
                        lst.push_back(a);
                        continue;
                    }
                    if (a - i > 2) {
                        return false;
                    }
                    if (a - i == step) {
                        if (step == 2) {
                            step = -1;
                            continue;
                        }
                        step += 1;
                        continue;
                    }
                    lst.push_back(a);
                }
                return isResolved(lst);
            }
            return false;
        }
        return true;
    }

    static void checkAllTiles() {
        if (allTiles[0] == 0) {
            int count = 0;
            for (int n : nums) {
                if (n >= 60) {
                    allTiles[count] = n;
                    count += 1;
                    continue;
                }
                for (int i = 1; i < 10; i++) {
                    allTiles[count] = n + i;
                    count += 1;
                }
            }
        }
    }

    static void checkDeckTiles() {
        if (deckTiles[0] == 0) {
            checkAllTiles();
            for (int i = 0; i < 4; i++) {
                copy(allTiles, end(allTiles), deckTiles + i * size(allTiles));
            }
        }
    }

public:
    static bool isHu(const vector<int>& numbers) {
        vector<int> eyes = getDuplicates(numbers, 2);

        for (int eye : eyes) {
            vector<int> ns = removeDuplicates(numbers, eye, 2);
            if (isResolved(ns)) return true;
        }

        return false;
    }

    static vector<int> listTin(const vector<int>& numbers) {
        vector<int> fullTiles = getDuplicates(numbers, 4);
        vector<int> tinTiles;

        checkAllTiles();
        for (int t : allTiles) {
            if (find(fullTiles.begin(), fullTiles.end(), t) != fullTiles.end()) {
                continue;
            }

            vector<int> ns(numbers);
            ns.push_back(t);
            sort(ns.begin(), ns.end());

            if (isHu(ns)) {
                tinTiles.push_back(t);
            }
        }
        return tinTiles;
    }

    static vector<int> tilesFromString(const string& tileString) {
        vector<int> tiles;
        int order = 0;
        for (size_t i = 0; i < tileString.length();) {
            int cplen = 1;
            if ((tileString[i] & 0xf8) == 0xf0) cplen = 4;
            else if ((tileString[i] & 0xf0) == 0xe0) cplen = 3;
            else if ((tileString[i] & 0xe0) == 0xc0) cplen = 2;
            if ((i + cplen) > tileString.length()) cplen = 1;

            const string c = tileString.substr(i, cplen);
            i += cplen;

            if (isdigit(c[0])) {
                order = c[0] - '0';
                continue;
            }
            if (c == " ") {
                order = 0;
                continue;
            }
            tiles.push_back(make(c, order));
        }
        sort(tiles.begin(), tiles.end());
        return tiles;
    }

    static string translate(const int& num) {
        return numberToString(num);
    }

    static string translate(const vector<int>& numbers) {
        string str;
        for (size_t i = 0; i < numbers.size(); i++) {
            int num = numbers[i];
            str += numberToString(num);
            if (i != numbers.size() - 1) str += " ";
        }
        return str;
    }

    static vector<int> draw(const int& num) {
        checkDeckTiles();
        random_device rd;
        shuffle(deckTiles, end(deckTiles), default_random_engine(rd()));
        vector<int> drawTiles(deckTiles, deckTiles + num);
        sort(drawTiles.begin(), drawTiles.end());
        return drawTiles;
    }
};

class Timer {
private:
    chrono::time_point<chrono::high_resolution_clock> startTime;
public:
    Timer() {
        startTime = chrono::high_resolution_clock::now();
    }
    void log(const string& message, const int& count=1) {
        chrono::duration<double> elapsed = chrono::high_resolution_clock::now() - startTime;
        cout << message << fixed << elapsed.count() / count << endl;
    }
};

void performanceTin(const int& count) {
    vector<int> tilesList[count];
    for (int i = 0; i < count; i++) {
        tilesList[i] = Tile::draw(16);
    }

    Timer timer = Timer();
    for (const vector<int>& tiles : tilesList) {
        Tile::listTin(tiles);
    }
    timer.log("Per tin time: ", count);
}

void findHuForm() {
    int count = 0;
    int th = 0;
    vector<int> tiles;
    Timer timer = Timer();

    do {
        tiles = Tile::draw(17);
        count += 1;
        th += 1;
        if (th == 10000) {
            th = 0;
            cout << count << endl;
        }
    } while (!Tile::isHu(tiles));

    cout << "Found form: " << Tile::translate(tiles) << endl;
    cout << "Draw count: " << count << endl;
    timer.log("Per draw & isHu time: ", count);
}

int main() {
    vector<int> tiles = Tile::tilesFromString("6萬 7萬 8萬 5條 6條 7條 9條 9條 4筒 5筒 6筒 7筒 8筒 9筒 中 中");
    cout << Tile::translate(Tile::listTin(tiles)) << endl;

    performanceTin(10000);

    findHuForm();

    return 0;
}

File: mahjong.lua
function find_index(arr, value)
    for k, v in ipairs(arr) do
        if (v == value) then
            return k
        end
    end
    return -1
end

function aprint(arr)
    local str = {}
    for _, v in ipairs(arr) do
        str[#str + 1] = v
    end
    print('['..table.concat(str, ', ')..']')
end

STRS = {'萬', '條', '筒', '東', '南', '西', '北', '中', '發', '白'}
NUMS = {0, 20, 40, 60, 65, 70, 75, 80, 85, 90}

function number_to_string(num)
    if num >= 60 then
        return STRS[find_index(NUMS, num)]
    end
    local order = num % 10
    return order..STRS[find_index(NUMS, num - order)]
end

function make(_type, order)
    local order = order or 0
    return NUMS[find_index(STRS, _type)] + order
end

function make_all(duplicates)
    duplicates = duplicates or 1
    local tiles = {}
    repeat
        for _, n in ipairs(NUMS) do
            if n >= 60 then
                table.insert(tiles, n)
                goto continue
            end
            for i = 1, 9 do
                tiles[#tiles + 1] = n + i
            end
            ::continue::
        end
        duplicates = duplicates - 1
    until duplicates == 0
    return tiles
end

ALL_TILES = make_all()

function get_duplicates(numbers, times)
    local last = numbers[1]
    local ocu = 1
    local result = {}
    for idx = 2, #numbers do
        local i = numbers[idx]
        if i == last then
            ocu = ocu + 1
            if ocu == times then
                result[#result + 1] = i
            end
        else
            ocu = 1
        end
        last = i
    end
    return result
end

function remove_duplicates(numbers, number, times)
    local result = {}
    for _, i in ipairs(numbers) do
        if i == number and times ~= 0 then
            times = times - 1
            goto continue
        end
        result[#result + 1] = i
        ::continue::
    end
    return result
end

function is_resolved(numbers)
    local is_ignore = false
    local ignore_times = 0

    for idx, i in ipairs(numbers) do
        if is_ignore and ignore_times ~= 0 then
            ignore_times = ignore_times - 1
            if ignore_times == 0 then
                is_ignore = false
            end
            goto continue
        end

        if idx + 2 > #numbers then
            return false
        end

        if i == numbers[idx + 2] then
            is_ignore = true
            ignore_times = 2
            goto continue
        end

        if i < 60 then
            local step = 1
            local lst = {}
            for aidx = idx + 1, #numbers do
                a = numbers[aidx]
                if step == -1 then
                    lst[#lst + 1] = a
                    goto acontinue
                end
                if a - i > 2 then
                    return false
                end
                if a - i == step then
                    if step == 2 then
                        step = -1
                    else
                        step = step + 1
                    end
                    goto acontinue
                end
                lst[#lst + 1] = a
                ::acontinue::
            end
            return is_resolved(lst)
        end

        if true then
            return false
        end

        ::continue::
    end
    return true
end

function is_hu(numbers)
    local eyes = get_duplicates(numbers, 2)
    for _, eye in ipairs(eyes) do
        local ns = remove_duplicates(numbers, eye, 2)
        if is_resolved(ns) then
            return true
        end
    end
    return false
end

function list_tin(numbers)
    local full_tiles = get_duplicates(numbers, 4)
    local tin_tiles = {}
    for _, t in ipairs(ALL_TILES) do
        if find_index(full_tiles, t) == -1 then
            local tiles = {t}
            table.move(numbers, 1, #numbers, 2, tiles)
            table.sort(tiles)
            if is_hu(tiles) then
                tin_tiles[#tin_tiles + 1] = t
            end
        end
    end
    return tin_tiles
end

function tiles_from_string(tile_string)
    local tiles = {}
    for s in string.gmatch(tile_string, '([^%s]+)') do
        if string.len(s) == 3 then
            tiles[#tiles + 1] = make(s)
        else
            local order = tonumber(string.sub(s, 1, 1))
            tiles[#tiles + 1] = make(string.sub(s, 2, 4), order)
        end
    end
    return tiles
end

function translate(num)
    if type(num) == 'number' then
        return number_to_string(num)
    end
    local str = {}
    for _, v in ipairs(num) do
        str[#str + 1] = number_to_string(v)
    end
    return table.concat(str, ' ')
end

DECK_TILES = make_all(4)

function draw(num)
    for i = #DECK_TILES, 2, -1 do
        local j = math.random(i)
        DECK_TILES[i], DECK_TILES[j] = DECK_TILES[j], DECK_TILES[i]
    end
    local tiles = table.move(DECK_TILES, 1, num, 1, {})
    table.sort(tiles)
    return tiles
end

set_time = 0
function timer(message, count)
    count = count or 1
    if set_time ~= 0 then
        print(
            string.format(
                message..' %.8f',
                (os.clock() - set_time) / count
            )
        )
        set_time = 0
    else
        set_time = os.clock()
    end
end

function performance_tin(count)
    local tiles_list = {}
    for i = 1, count do
        tiles_list[#tiles_list + 1] = draw(16)
    end
    timer()
    for _, tiles in ipairs(tiles_list) do
        list_tin(tiles)
    end
    timer('Per tin time', count)
end

function find_hu_form()
    local count = 0
    local th = 0
    local tiles
    timer()
    repeat
        tiles = draw(17)
        count = count + 1
        th = th + 1
        if th == 10000 then
            th = 0
            print(count)
        end
    until is_hu(tiles)

    print('Found form: '..translate(tiles))
    print('Draw count: '..count)
    timer('Per draw & isHu time: ', count)
end
    

test_tiles = tiles_from_string('6萬 7萬 8萬 5條 6條 7條 9條 9條 4筒 5筒 6筒 7筒 8筒 9筒 中 中')
print(translate(list_tin(test_tiles)))

performance_tin(10000)

find_hu_form()
